site/documentation/content/api/device-connection/index.md

```diff
@@ -141,9 +142,11 @@ The response message's *status* property may contain the following codes:
 | Code  | Description |
 | :---- | :---------- |
 | *204* | OK, the adapter instance for the device has been updated. |
-| *400* | Bad Request, the command-handling adapter instance has not been updated due to invalid or missing data in the request. |
+| *400* | Bad Request, the command-handling adapter instance has not been set or updated due to invalid or missing data in the request. |
+| *404* | Not Found, entry to be updated was not found. This applies to requests with *only_update* set to `true`. |
```

Comment:
FMPOV this should be returned if no entry has been found for the given device ID, regardless of the value of *update_only*.
We should then expicitly list *412* as the value to be returned for *update_only* = `true` and a non-matching orig-value. I do not see why the registry implementation should return 404 or 412 in that case, given that we expect the registry to explicitly support this property anyway ...

Reply:
> FMPOV this should be returned if no entry has been found for the given device ID, regardless of the value of update_only.

But, whether an entry for the given device ID is found or not, only matters if `update_only` is `true` (if `update_only` is false, a new mapping entry with the device ID will just be created).

> We should then expicitly list *412* as the value to be returned for *update_only* = `true` and a non-matching orig-value. I do not see why the registry implementation should return 404 or 412 in that case, given that we expect the registry to explicitly support this property anyway ...

Two things made me go for choosing *404* as the default for both the key-not-found and adapter-instance-not-matching cases here:
- the `removeCommandHandlingAdapterInstance` command also requires the adapterInstanceId to match. For that command we have defined 404 as the default if no matching entry was found, and 412 as an optional alternative to 404.
- a straightforward implementation might use a `replace(key, oldValue, newValue)` method here. With that, there is no distinction, whether an entry for the key wasn't found or whether the `oldValue` didn't match. In fact for the cache based implementation, I've used the corresponding `cache.replaceAsync()` method. (That could of course be changed to use `cache.computeIfAbsentAsync` where a distinction between return values can be made then.)

Therefore, for the sake of keeping in sync with the other API methods and for making it easier for implementations (not requiring the key-not-found vs adapter-instance-not-matching distinction, which also isn't needed in the code using the API), I would rather keep the API definition as is.

Reply:
> But, whether an entry for the given device ID is found or not, only matters if update_only is true (if update_only is false, a new mapping entry with the device ID will just be created).

true, based on that it would probably make sense to never return 404 but to return 412 if *update_only* = `true` and either no entry for the given device ID exists at all or the expected value doesn't match.

> the removeCommandHandlingAdapterInstance command also requires the adapterInstanceId to match. For that command we have defined 404 as the default if no matching entry was found, and 412 as an optional alternative to 404.

maybe we should change that as well. I don't really like the fact that a client needs to support both status codes but still doesn't really know what the problem is ...

Reply:
I would be fine with always returning 412 (and not having the distinction between the two cases).

To be consistent, I agree that `removeCommandHandlingAdapterInstance` should then be changed as well. I think, regarding compatibility, we can do this in 1.3, since "412" is already defined there as a possible return value.

Another thing would be changing the 404 status code for the optional tenant check in the set/remove operations. I would like to change that to 412 as well. But since `setLastKnownGateway` doesn't define 412 as possible return code as of now, I think we can only do this in Hono 2.0.
Therefore I would keep the 404 as the common status code for the tenant check in all methods for now.

Reply:
I've updated the documentation accordingly.
(Changes concerning `removeCommandHandlingAdapterInstance` will go in a separate PR then.)